src/GraphQL/Validation/DocumentValidator.cs

```diff
@@ -107,27 +128,43 @@ public class DocumentValidator : IDocumentValidator
                 }
                 else
                 {
-                    var awaitedVisitors = rules.Select(x => x.ValidateAsync(context)).Where(x => x != null);
+                    var awaitedVisitors = rules.Select(x =>
+                    {
+                        if (x is IVariableValidation validation)
+                        {
+                            var variableVisitor = validation.GetVisitor(context);
+                            if (variableVisitor != null)
+                                (variableVisitors ??= new List<IVariableVisitor>()).Add(variableVisitor);
+                        }
+                        return x.ValidateAsync(context);
+                    }).Where(x => x != null);
                     visitors = (await Task.WhenAll(awaitedVisitors)).ToList();
                 }
 
                 visitors.Insert(0, context.TypeInfo);
 
                 new BasicVisitor(visitors).Visit(document, context);
 
-                success = !context.HasErrors;
-                return success
-                    ? SuccessfullyValidatedResult.Instance
-                    : new ValidationResult(context.Errors);
+                var variables = context.GetVariableValues(schema, variableDefinitions, inputs ?? Inputs.Empty, variableVisitors == null ? null : new CompositeVariableVisitor(variableVisitors));
```

Comment:
I know performance along this path isn't critical, but it would be nice if it skipped initialization of `CompositeVariableVisitor` when there was only a single `variableVisitor`

Reply:
Done